In this notebook, you'll see how to connect to a Postgres database using the sqlalchemy library.

For this notebook, you'll need both the `sqlalchemy` and `psycopg2` libraries installed.

In [1]:
from sqlalchemy import create_engine

First, we need to create a connection string. The format is

 ```<dialect(+driver)>://<username>:<password>@<hostname>:<port>/<database>```

To connect to the Lahman baseball database, you can use the following connection string.

In [2]:
database_name = 'scooters'    # Fill this in with your scooter database name

connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"

Now, we need to create an engine and use it to connect.

In [3]:
engine = create_engine(connection_string)

Now, we can create our query and pass it into the `.query()` method.

In [4]:
# Look at difference in run time for this:
query = '''
SELECT latitude
FROM scooters;
'''

result = engine.execute(query)

In [5]:
# Vs this:
query = '''
SELECT COUNT(latitude)
FROM scooters;
'''

result = engine.execute(query)

You can then fetch the results as tuples using either `fetchone` or `fetchall`:

In [6]:
result.fetchone()

(73414043,)

On the other hand, sqlalchemy plays nicely with pandas.

In [7]:
import pandas as pd

In [8]:
lat = pd.read_sql(query, con = engine)
lat.head()

count
0  73414043

Are there any null values in any columns in either table?

In [9]:
query = '''

SELECT COUNT(*)
FROM scooters
WHERE pubdatetime IS NULL
'''

In [10]:
pubdatetimenulls = pd.read_sql(query, con=engine)
print(pubdatetimenulls)

   count
0      0


In [11]:
query = '''

SELECT COUNT(*)
FROM scooters
WHERE chargelevel IS NULL

'''

In [12]:
chargelevel_nulls = pd.read_sql(query, con=engine)
print(chargelevel_nulls)

   count
0    770


In [13]:
query= '''
SELECT SUM(CASE WHEN chargelevel is null THEN 1 ELSE 0 END) 
AS charge_nulls 
FROM scooters
'''

In [14]:
charge_nulls = pd.read_sql(query, con=engine)
print(charge_nulls)

   charge_nulls
0           770


In [15]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE enddate IS NULL

'''


In [16]:
enddate_nulls = pd.read_sql(query, con=engine)
print(enddate_nulls)

   count
0      0


In [17]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE endtime IS NULL

'''

In [18]:
endtime_nulls = pd.read_sql(query, con=engine)
print(endtime_nulls)

   count
0      0


In [19]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE startlatitude IS NULL

'''

In [20]:
startlatitude_nulls = pd.read_sql(query, con=engine)
print(startlatitude_nulls)

   count
0      0


In [21]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE startlongitude IS NULL

'''

In [22]:
startlongitude_nulls = pd.read_sql(query, con=engine)
print(startlongitude_nulls)

   count
0      0


In [23]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE endlatitude IS NULL

'''

In [24]:
endlatitude_nulls = pd.read_sql(query, con=engine)
print(endlatitude_nulls)

   count
0      0


In [25]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE endlongitude IS NULL

'''

In [26]:
endlongitude_nulls = pd.read_sql(query, con=engine)
print(endlongitude_nulls)

   count
0      0


In [27]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE triproute IS NULL

'''

In [28]:
triproute_nulls = pd.read_sql(query, con=engine)
print(triproute_nulls)

   count
0      0


In [29]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE create_dt IS NULL

'''

In [30]:
create_dt_nulls = pd.read_sql(query, con=engine)
print(create_dt_nulls)

   count
0      0


In [31]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE sumdid IS NULL

'''

In [32]:
sumdid_nulls = pd.read_sql(query, con=engine)
print(sumdid_nulls)

   count
0      0


In [33]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE tripdistance IS NULL

'''

In [34]:
tripdistance_nulls = pd.read_sql(query, con=engine)
print(tripdistance_nulls)

   count
0      0


Answer: 770 nulls on scooters chargelevel


What date range is represented in each of the date columns? Investigate any values that seem odd.

In [35]:
query = '''SELECT MIN(pubdatetime)
FROM scooters
'''

In [36]:
min_scooter_pubdatetime = pd.read_sql(query, con=engine)
print(min_scooter_pubdatetime)

                      min
0 2019-05-01 00:01:41.247


In [37]:
query = '''SELECT MAX(pubdatetime)
FROM scooters
'''

In [38]:
max_scooter_pubdatetime = pd.read_sql(query, con=engine)
print(max_scooter_pubdatetime)

                  max
0 2019-07-31 23:59:57


In [39]:
query = '''SELECT MIN(pubtimestamp)
FROM trips
'''

In [40]:
min_trips_pubtimestamp = pd.read_sql(query, con=engine)
print(min_trips_pubtimestamp)

                      min
0 2019-05-01 00:00:55.423


In [41]:
query = '''SELECT MAX(pubtimestamp)
FROM trips
'''

In [42]:
max_trips_pubtimestamp = pd.read_sql(query, con=engine)
print(max_trips_pubtimestamp)

                  max
0 2019-08-01 07:04:00


In [43]:
query = '''SELECT DISTINCT sumdgroup
FROM scooters
'''


In [44]:
distinct_sumdgroup = pd.read_sql(query, con=engine)
print(distinct_sumdgroup)

  sumdgroup
0   bicycle
1   scooter
2   Scooter


What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?

In [45]:
query = '''SELECT MIN(startlatitude)
FROM trips
'''

In [46]:
min_start_latitude = pd.read_sql(query, con=engine)
print(min_start_latitude)

       min
0  35.8532


In [47]:
query = '''SELECT MAX(startlatitude)
FROM trips
'''

In [48]:
max_start_latitude = pd.read_sql(query, con=engine)
print(max_start_latitude)

         max
0  36.300029


In [49]:
query = '''SELECT MIN(startlongitude)
FROM trips
'''

In [50]:
min_start_longitude = pd.read_sql(query, con=engine)
print(min_start_longitude)

         min
0 -86.918008


In [51]:
query = '''SELECT MAX(startlongitude)
FROM trips
'''

In [52]:
max_start_longitude = pd.read_sql(query, con=engine)
print(max_start_longitude)

       max
0 -86.3662


In [53]:
query = '''SELECT MAX(tripduration)
FROM trips
'''

In [54]:
max_trip_duration = pd.read_sql(query, con=engine)
print(max_trip_duration)

        max
0  512619.0


In [55]:
query = '''SELECT MIN(tripduration)
FROM trips
'''

In [56]:
min_trip_duration = pd.read_sql(query, con=engine)
print(min_trip_duration)

         min
0 -19.358267


1. During this period, seven companies offered scooters. How many scooters did each company have in this time frame? Did the number for each company change over time? Did scooter usage vary by company?

In [59]:
query = '''SELECT DISTINCT companyname
FROM scooters'''

In [60]:
scooter_companies = pd.read_sql(query, con=engine)
print(scooter_companies)

  companyname
0        Bird
1        Bolt
2      Gotcha
3        Jump
4        Lime
5        Lyft
6        Spin


In [61]:
import pandas as pd

In [62]:
import numpy as np

In [75]:
import matplotlib as plt

In [64]:
query = '''SELECT COUNT(sumdid)
FROM scooters
WHERE companyname = 'Bird'
'''

In [65]:
bird_count = pd.read_sql(query, con=engine)
print(bird_count)

      count
0  12251590


In [68]:
query = '''SELECT COUNT(sumdid)
FROM scooters
WHERE companyname = 'Bolt'
'''

In [69]:
bolt_count = pd.read_sql(query, con=engine)
print(bolt_count)

     count
0  3477198


In [71]:
query = '''
SELECT DISTINCT(sumdid),
companyname
FROM scooters;
'''

In [73]:
company_scooters = pd.read_sql(query, con = engine)
company_scooters.head()

sumdid companyname
0  Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619        Bolt
1  Powered-01a24436-0315-e1bb-7ce0-d081d05dff7d        Bolt
2  Powered-03be23ca-d43b-222f-be54-e44b5b4690df        Bolt
3  Powered-046201fb-6532-1f37-6334-3612fb1e61f7        Bolt
4  Powered-0479bb84-afbd-0426-f1c4-df628542a88c        Bolt

In [74]:
df2 = company_scooters.groupby(['companyname'])['companyname'].count()
print(df2)

companyname
Bird      3860
Bolt       360
Gotcha     224
Jump      1210
Lime      1824
Lyft      1735
Spin       805
Name: companyname, dtype: int64


In [86]:
query = '''
SELECT sumdid,
companyname, MIN(pubdatetime)
FROM scooters
GROUP BY sumdid, companyname;
'''

company_scooters_time = pd.read_sql(query, con = engine)
company_scooters_time.head()

sumdid companyname   
0  Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619        Bolt  \
1  Powered-01a24436-0315-e1bb-7ce0-d081d05dff7d        Bolt   
2  Powered-03be23ca-d43b-222f-be54-e44b5b4690df        Bolt   
3  Powered-046201fb-6532-1f37-6334-3612fb1e61f7        Bolt   
4  Powered-0479bb84-afbd-0426-f1c4-df628542a88c        Bolt   

                  min  
0 2019-05-24 00:04:42  
1 2019-05-24 00:04:42  
2 2019-05-24 00:04:42  
3 2019-05-28 20:53:56  
4 2019-05-24 00:04:42

In [77]:
import seaborn as sb

In [83]:
query = '''SELECT * 
FROM scooters
LIMIT 5'''

scooters = pd.read_sql(query, con=engine)
print(scooters)

          pubdatetime  latitude  longitude   
0 2019-07-01 23:08:43   36.1200   -86.7535  \
1 2019-07-01 23:08:43   36.1201   -86.7533   
2 2019-07-01 23:08:43   36.1203   -86.7532   
3 2019-07-01 23:08:43   36.1200   -86.7532   
4 2019-07-01 23:08:43   36.1199   -86.7535   

                                        sumdid sumdtype  chargelevel   
0  Powered61f29e0c-a351-55b4-879f-b2be89221570  Powered          1.0  \
1  Powerede273ae04-088b-5739-bbda-ef6b6c03a190  Powered          1.0   
2  Powered30ef1232-1eed-57fb-b558-07356d1c4eea  Powered         69.0   
3  Poweredd2ea1e00-725b-506f-b6ba-2e93e64010e0  Powered          3.0   
4  Powered8a441602-0855-592d-98ac-2276e8ae48dd  Powered          0.0   

  sumdgroup  costpermin companyname  
0   scooter        0.06        Jump  
1   scooter        0.06        Jump  
2   scooter        0.06        Jump  
3   scooter        0.06        Jump  
4   scooter        0.06        Jump  


In [84]:
query = '''SELECT *
FROM trips
LIMIT 5'''

trips = pd.read_sql(query, con=engine)
print(trips)

             pubtimestamp companyname triprecordnum         sumdid   
0 2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL  \
1 2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2 2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3 2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4 2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance   startdate        starttime     enddate   
0        3.0000     958.00528  2019-05-01  00:00:20.460000  2019-05-01  \
1        1.7156    1371.39112  2019-05-01  00:01:50.090000  2019-05-01   
2        3.0000    2296.58800  2019-05-01  00:03:47.363333  2019-05-01   
3        3.0000    1200.78744  2019-05-01  00:04:21.386666  2019-05-01   
4        2.0000     351.04988  2019-05-01  00:04:27.796666  2019-05-01   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude   
0  00:02:52.346666       36.15710       -86.80360     

In [87]:
company_scooters_time['min'] = pd.to_datetime(company_scooters_time['min'])

In [88]:
company_scooters_time.groupby([company_scooters_time['companyname'], company_scooters_time['min'].dt.date]).count()

sumdid   min
companyname min                     
Bird        2019-05-01    1545  1545
            2019-05-02     164   164
            2019-05-03      55    55
            2019-05-04      29    29
            2019-05-05       6     6
...                        ...   ...
Spin        2019-07-14       1     1
            2019-07-17       1     1
            2019-07-19       1     1
            2019-07-22       1     1
            2019-07-26       2     2

[317 rows x 2 columns]

In [100]:
query = '''SELECT COUNT(DISTINCT sumdid), EXTRACT(MONTH FROM pubdatetime) as month, companyname 
FROM scooters
GROUP BY month, companyname;
'''

scooters_by_month = pd.read_sql(query, con=engine)
print(scooters_by_month)

    count  month companyname
0    3064    5.0        Bird
1     346    5.0        Bolt
2     223    5.0      Gotcha
3     976    5.0        Jump
4    1632    5.0        Lime
5    1376    5.0        Lyft
6     687    5.0        Spin
7    2910    6.0        Bird
8     333    6.0        Bolt
9     223    6.0      Gotcha
10   1036    6.0        Jump
11   1461    6.0        Lime
12   1152    6.0        Lyft
13    713    6.0        Spin
14   2583    7.0        Bird
15    276    7.0        Bolt
16    224    7.0      Gotcha
17   1206    7.0        Jump
18    903    7.0        Lime
19   1026    7.0        Lyft
20    570    7.0        Spin


In [102]:
import seaborn as sns

In [103]:
sns.lineplot(data = scoot_df, x='date', y='cumulative_count', hue='companyname')

NameError: name 'scoot_df' is not defined